auth/gcloud/aio/auth/token.py

```diff
@@ -267,18 +326,158 @@ async def acquire_access_token(self, timeout: int = 10) -> None:
         else:
             raise Exception(f'unsupported token type {self.token_type}')
 
+        return resp
+
+
+class IapToken(BaseToken):
+    """An OpenID Connect ID token for a single IAP-secured service."""
+
+    def __init__(
+        self, app_uri: str,
+        service_file: Optional[Union[str, IO[AnyStr]]] = None,
+        session: Optional[Session] = None,
+        service_account: Optional[str] = None,
+    ) -> None:
+        super().__init__(service_file=service_file, session=session)
+
+        self.app_uri = app_uri
+        self.service_account = service_account
+
+        if (self.token_type == Type.AUTHORIZED_USER
+                and not self.service_account):
+            raise Exception(
+                'service account name must be provided when token type is '
+                'authorized user',
+            )
+
+    async def _get_iap_client_id(self, *, timeout: int) -> str:
+        """
+        Fetch the IAP client ID from the service URI.
+
+        If not logged in already, then we parse the OAuth redirect location to
+        get the client ID. The redirect location is a header of the form:
+
+            https://accounts.google.com/o/oauth2/v2/auth?client_id=<id>&...
+
+        For more details, see the GCP docs for programmatic IAP access:
+        https://cloud.google.com/iap/docs/authentication-howto
+        """
+        resp = await self.session.head(self.app_uri, timeout=timeout,
+                                       allow_redirects=False)
+
+        redirect_location = resp.headers.get('location')
+        if not redirect_location:
+            raise Exception(f'No redirect location for service {self.app_uri},'
+                            ' is it secured with IAP?')
+
+        parsed_uri = urlparse(redirect_location)
+        query = parse_qs(parsed_uri.query)
+        client_id: str = query.get('client_id', [''])[0]
+        if not client_id:
+            raise Exception(f'No client ID found for service {self.app_uri},'
+                            ' is it secured with IAP?')
+        return client_id
+
+    async def _refresh_authorized_user(
+        self, iap_client_id: str,
+        timeout: int,
+    ) -> TokenResponse:
+        """
+        Fetch IAP ID token by impersonating a service account.
+
+        https://cloud.google.com/iap/docs/authentication-howto#obtaining_an_oidc_token_in_all_other_cases
+        """
+        async def get_access_token(timeout: int) -> str:
```

Comment:
nit: consider de-nesting this method to allow us to more easily test it on its own.

Reply:
I am with Kevin here, unless there are compelling reasons not to.

Reply:
I'll go with the majority on this one and de-nest this one, but I personally find inline functions to be great for readability when a chunk of code doesn't deserve its own method but represents a block of functionality which serves a specific purpose. I.e. it can be represented as an input/output contract via a function.